js/serializers/scormSuspendDataSerializer.js

```diff
@@ -1,691 +1,1129 @@
+        bin.unshift(valueType.binType);
+      }
+      logStats && this.printLog();
+      return bin;
+    }
 
-        return base64;
-      },
-      "base16": function ConverterBinaryToBase16(bin) {
-        var paddingLength = 0;
-        if (bin.length % 4) paddingLength = 4 - (bin.length % 4);
-        binPaddingLen = NumberToBinary(paddingLength, 4);
-        binPadding = Padding.addLeft("", paddingLength);
-        bin = binPaddingLen + binPadding + bin;
-
-        var binLength = bin.length;
-        var hex = "";
-        for (var b = 0; b < 10000; b++) {
-          if (b*4 >= binLength) break;
-
-          var block = bin.substr(b*4,4);
-          hex += parseInt(block, 2).toString(16);
+    /**
+     * Convert the string binary representation back into an array, boolean or number.
+     * @param {string} bin A string binary representation of data from valueToBin
+     * @param {string} [typeName] To force an internal data type for testing
+     * @returns {number|boolean|Array}
+     */
+    valueFromBin(bin, typeName = null) {
+      bin = _.flatten(bin).join('');
+      let valueType;
+      if (typeof typeName === 'string') {
+        valueType = findValueTypeFromName(typeName);
+        if (!valueType) {
+          throw new Error(`Could not find value type from name "${typeName}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
+        }
+      } else {
+        [valueType, bin]= shiftValueTypeFromBin(bin);
+        if (!valueType || !(valueType instanceof AbstractValueType)) {
+          throw new Error(`Could not find value type from value "${value}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
         }
-        return hex;
-      },
-      "double": function ConverterBinaryToDouble(bin, wrap) {
-        var typeDef = DataType("double");
-
-        if (wrap !== false)
-            bin = Converter._variableUnwrapLength(bin);
-
-        var isMinus = bin.substr(0 ,1) == 1;
-
-        var exponentByte = parseInt("0" + bin.substr(1, 7), 2);
-        var baseLong = parseInt( bin.substr(8, bin.length), 2);
-
-        var dbl = parseFloat(baseLong+"E-"+exponentByte, 10);
-        if (isMinus) dbl = dbl * -1;
-
-        return dbl;
-      },
-      "long": function ConverterBinaryToLong(bin) {
-        return parseInt(bin.substr(0, 32), 2);
-      },
-      "short": function ConverterBinaryToShort(bin) {
-        return parseInt(bin.substr(0, 16), 2);
-      },
-      "byte": function ConverterBinaryToByte(bin) {
-        return parseInt(bin.substr(0, 8), 2);
-      },
-      "half": function ConverterBinaryToHalf(bin) {
-        return parseInt(bin.substr(0, 4), 2);
-      },
-      "boolean": function ConverterBinaryToBoolean(bin) {
-        return bin.substr(0,1) == "1" ? true: false;
-      },
-      "number": function ConverterBinaryToNumber(bin) {
-        return parseInt(bin, 2);
       }
+      let value;
+      [value] = valueType.shiftValueFromBin(bin);
```

Comment:
Anything obvious preventing this?
```suggestion
      const [ value ] = valueType.shiftValueFromBin(bin);
```

Reply:
yup